metagpt/environment/werewolf_env/werewolf_ext_env.py

```diff
@@ -0,0 +1,274 @@
+    16: {
+        "content": """Living players: {living_players}, now freely talk about the current situation based on your observation and
+                    reflection with a few sentences. Decide whether to reveal your identity based on your reflection.""",
+        "send_to": "",  # send to all to speak in daytime
+        "restricted_to": "",
+    },
+    17: {
+        "content": """Now vote and tell me who you think is the werewolf. Don’t mention your role.
+                    You only choose one from the following living options please:
+                    {living_players}. Say ONLY: I vote to eliminate ...""",
+        "send_to": "",
+        "restricted_to": "",
+    },
+    18: {"content": """{player_current_dead} was eliminated.""", "send_to": "Moderator", "restricted_to": ""},
+}
+
+
+class WerewolfExtEnv(ExtEnv):
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    roles_state: dict[str, RoleState] = Field(default=dict(), description="the role's current state by role_name")
+
+    step_idx: int = Field(default=0)  # the current step of current round
+    eval_step_idx: int = Field(default=0)
+    per_round_steps: int = Field(default=len(STEP_INSTRUCTIONS))
+
+    # game global states
+    game_setup: str = Field(default="", description="game setup including role and its num")
+    living_players: list[str] = Field(default=[])
+    werewolf_players: list[str] = Field(default=[])
+    villager_players: list[str] = Field(default=[])
+    special_role_players: list[str] = Field(default=[])
+    winner: Optional[str] = Field(default=None)
+    win_reason: Optional[str] = Field(default=None)
+    witch_poison_left: int = Field(default=1)
+    witch_antidote_left: int = Field(default=1)
+
+    # game current round states, a round is from closing your eyes to the next time you close your eyes
+    player_hunted: Optional[str] = Field(default=None)
+    player_protected: Optional[str] = Field(default=None)
+    is_hunted_player_saved: bool = Field(default=False)
+    player_poisoned: Optional[str] = Field(default=None)
+    player_current_dead: list[str] = Field(default=[])
+
+    def parse_game_setup(self, game_setup: str):
+        self.game_setup = game_setup
+        self.living_players = re.findall(r"Player[0-9]+", game_setup)
+        self.werewolf_players = re.findall(r"Player[0-9]+: Werewolf", game_setup)
+        self.werewolf_players = [p.replace(": Werewolf", "") for p in self.werewolf_players]
+        self.villager_players = re.findall(r"Player[0-9]+: Villager", game_setup)
+        self.villager_players = [p.replace(": Villager", "") for p in self.villager_players]
+        self.special_role_players = [
+            p for p in self.living_players if p not in self.werewolf_players + self.villager_players
+        ]
+
+        # init role state
+        self.roles_state = {player_name: RoleState.ALIVE for player_name in self.living_players}
+
+    @mark_as_readable
+    def init_game_setup(
```

Comment:
Should this method be exposed?

Reply:
It provides game_setup and role instances from the env, so mark it as a readable one.